### ISSUE TITLE
Classification recognition rate issue

### ISSUE THREAD

**Issue by monkeycc** [ultralytics/ultralytics#20645]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

```python
    model = YOLO( "./yolo11s-cls.pt") 

    results = model.train(data=  "./yolo", imgsz=512 , epochs=100, batch = 60 ,   degrees=0.0, translate=0.0, scale=0.0, shear=0.0, perspective=0.0,
                      hsv_h=0.0,hsv_s=0.0, hsv_v=0.0, flipud=0.0, fliplr=0.0, bgr=0.0, mosaic=0.0, mixup=0.0, copy_paste=0.0, copy_paste_mode='', auto_augment='',
                      erasing=0.0, crop_fraction=1.0)  
```

![Image](https://github.com/user-attachments/assets/4f1ebfa3-33da-4749-8454-06b12df07833)

For the angle
The recognition rate is very poor

Request assistance

**Comment by UltralyticsAssistant** [comment]
👋 Hello @monkeycc, thank you for reaching out and sharing your experience with Ultralytics 🚀! This is an automated response to help you get started while an Ultralytics engineer reviews your issue.

We recommend visiting the [Docs](https://docs.ultralytics.com/) for detailed guides and checking out [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples, which may address common classification and training questions.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) (MRE) including a sample of your data and the exact code used. This helps us debug more efficiently.

For training performance or recognition rate ❓ questions, please share additional details such as:
- Dataset samples (images and labels)
- Complete training logs
- Your data.yaml configuration
- Clarification on how the "angle" is affecting results

You can also connect with the Ultralytics community in real-time on [Discord](https://discord.com/invite/ultralytics) 🎧, join discussions on [Discourse](https://community.ultralytics.com/), or share insights on our [Subreddit](https://reddit.com/r/Ultralytics).

## Upgrade

Please upgrade to the latest `ultralytics` package and ensure all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) are met in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to check if the issue persists:

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in these verified environments (with dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

An Ultralytics engineer will review your issue and assist you further soon. Thank you for your patience and for helping improve Ultralytics!

**Comment by glenn-jocher** [comment]
Based on your training configuration, I can see why you're experiencing poor recognition rates for objects at different angles. You've disabled all data augmentation (`degrees=0.0, translate=0.0, scale=0.0...`), which is likely limiting your model's ability to generalize to different orientations and variations.

For objects at different angles, I'd recommend:

```python
model.train(data="./yolo", imgsz=512, epochs=100, batch=60,
            degrees=45.0,  # Enable rotation augmentation
            fliplr=0.5,    # Add horizontal flips
            scale=0.5      # Add scaling variation
            # Keep other augmentations as needed
           )
```

Since your dataset contains objects with different orientations and missing corners, these augmentations will help the model learn invariance to these variations. Consider using `perspective` and `shear` parameters as well if you need the model to recognize objects from different viewpoints.

For more detailed guidance on model evaluation and fine-tuning, you can refer to our [performance metrics guide](https://docs.ultralytics.com/guides/yolo-performance-metrics/) to understand how to interpret and improve your results.

**Comment by monkeycc** [comment]
![Image](https://github.com/user-attachments/assets/d3d07597-d8fe-4f04-975d-6e825a619118)
---
![Image](https://github.com/user-attachments/assets/7d649915-a3da-419c-89a7-aa5065fe8330)
---
![Image](https://github.com/user-attachments/assets/508d64f2-7be7-417a-993f-0cefc3d1b699)
---

train_batch0.jpg
train_batch1.jpg
train_batch2.jpg

```python
(flipud=0.0, fliplr=0.0, mosaic=0.0, mixup=0.0, copy_paste=0.0, copy_paste_mode='', 
                      erasing=0.0, shear=0.0,translate=0.0,auto_augment=None,degrees=0.0)
```

I found that my image will be cropped
I need not to crop or move, but I found that it doesn't have much effect

Because flipping can lead to incorrect direction determination
Moving and cropping can lead to issues with recognition rate,
I just need the complete image, changing the color and brightness, and the tilt angle

But I found that the configuration file is useless and not effective

Why is it in train_match0 train_match1 train_match2
I see incomplete images
In theory, it should be a complete thumbnail

**Comment by glenn-jocher** [comment]
The cropping you're seeing is due to YOLO classification using `torchvision.transforms.RandomResizedCrop` by default, which crops random portions of images regardless of your augmentation settings. To preserve complete images while only applying color/brightness and rotation changes, you'll need to customize the transforms:

```python
import torch
import torchvision.transforms as T
from ultralytics import YOLO
from ultralytics.data.dataset import ClassificationDataset
from ultralytics.models.yolo.classify import ClassificationTrainer

class CustomDataset(ClassificationDataset):
    def __init__(self, root: str, args, augment: bool = False, prefix: str = ""):
        super().__init__(root, args, augment, prefix)
        train_transforms = T.Compose([
            T.Resize((args.imgsz, args.imgsz)),  # Resize without cropping
            T.RandomRotation(degrees=args.degrees),  # Only rotation
            T.ColorJitter(brightness=args.hsv_v, contrast=args.hsv_v, 
                         saturation=args.hsv_s, hue=args.hsv_h),  # Color changes
            T.ToTensor(),
            T.Normalize(mean=torch.tensor(0), std=torch.tensor(1))
        ])
        self.torch_transforms = train_transforms if augment else T.Compose([
            T.Resize((args.imgsz, args.imgsz)), T.ToTensor(),
            T.Normalize(mean=torch.tensor(0), std=torch.tensor(1))
        ])

class CustomTrainer(ClassificationTrainer):
    def build_dataset(self, img_path: str, mode: str = "train", batch=None):
        return CustomDataset(root=img_path, args=self.args, augment=mode == "train", prefix=mode)

model = YOLO("./yolo11s-cls.pt")
model.train(data="./yolo", trainer=CustomTrainer, imgsz=512, epochs=100, batch=60,
            degrees=30.0, hsv_h=0.02, hsv_s=0.5, hsv_v=0.5)
```

This replaces `RandomResizedCrop` with `Resize` to preserve complete images while applying only the augmentations you need.

**Comment by monkeycc** [comment]
Thank you.
effective

Another question is
`degrees=30.0`

![Image](https://github.com/user-attachments/assets/d6653f2c-d040-4fa5-b7f3-d2637c31412a)

Rotate 30 °
Is it a safe rotation
Will it cause the corners to rotate outside the image,
Causing incomplete image display
Thereby affecting the recognition rate

**Comment by glenn-jocher** [comment]
You raise an excellent point about rotation safety. With `degrees=30.0`, corners can indeed rotate outside the image boundaries, potentially cropping important parts of your objects and affecting recognition.

To avoid this issue, you have two options:

```python
# Option 1: Use smaller, safer rotation angles
T.RandomRotation(degrees=15.0)  # Reduces risk of corner cropping

# Option 2: Expand canvas to preserve entire rotated image
T.RandomRotation(degrees=30.0, expand=True)  # Keeps full image but changes aspect ratio
```

For your use case where you need complete images, I'd recommend starting with `degrees=15.0` as this provides rotation variation while minimizing the risk of cutting off important object features. You can gradually increase the angle if needed based on your results.

The `expand=True` option preserves the entire rotated image but changes the aspect ratio, which might not be ideal for classification training where consistent input sizes are preferred.
